ArduCopter/inertia.cpp

```diff
@@ -17,12 +17,15 @@ void Copter::read_inertia()
         return;
     }
 
-    if (ahrs.home_is_set()) {
-        current_loc.set_alt_cm(inertial_nav.get_altitude(),
-                               Location::AltFrame::ABOVE_HOME);
+    // current_loc.alt is alt-above-home, atomically converted from inertial nav's alt-above-ekf-origin
+    const int32_t alt_above_origin_cm = inertial_nav.get_altitude();
+    int32_t alt_above_home_cm;
+    loc.set_alt_cm(alt_above_origin_cm, Location::AltFrame::ABOVE_ORIGIN);
+    if (loc.get_alt_cm(Location::AltFrame::ABOVE_HOME, alt_above_home_cm)) {
+        current_loc.set_alt_cm(alt_above_home_cm, Location::AltFrame::ABOVE_HOME);
     } else {
-        // without home use alt above the EKF origin
-        current_loc.set_alt_cm(inertial_nav.get_altitude(),
-                               Location::AltFrame::ABOVE_ORIGIN);
+        // if home has not been set yet we treat alt-above-origin as alt-above-home
+        // current_loc's altitude frame is always ABOVE_HOME
+        current_loc.set_alt_cm(alt_above_origin_cm, Location::AltFrame::ABOVE_HOME);
     }
```

Comment:
I wouldn't use all of this, all we need is:

```cpp
current_loc.set_alt_cm(alt_above_origin_cm, Location::AltFrame::ABOVE_ORIGIN);

if (!current_loc.change_alt_frame(Location::AltFrame::ABOVE_HOME)) {
    // if home has not been set yet we treat alt-above-origin as alt-above-home
    current_loc.set_alt_cm(alt_above_origin_cm, Location::AltFrame::ABOVE_HOME);
}
```

Or more explicit:

```cpp
current_loc.set_alt_cm(alt_above_origin_cm, Location::AltFrame::ABOVE_ORIGIN);
bool changed_to_home_frame = current_loc.change_alt_frame(Location::AltFrame::ABOVE_HOME);

if (!changed_to_home_frame) {
    // if home has not been set yet we treat alt-above-origin as alt-above-home
    current_loc.set_alt_cm(alt_above_origin_cm, Location::AltFrame::ABOVE_HOME);
}
```

In my opinion the if part isn't even needed, but see comment below.

Reply:
Thanks very much for the review.  I was worried of separate threads trying to use current_loc.alt during the instant where it's frame is still alt-above-origin but after a review there don't seem to be any threads that do this so I've changed it to the shorter code you've suggested.